# Process data

In [1]:
import os

In [2]:
data_dir = os.path.join('..','resources','data')

os.listdir(data_dir)

['05-2019.csv',
 '06-2019.csv',
 '07-2019.csv',
 '08-2019.csv',
 '09-2019.csv',
 '10-2019.csv',
 '11-2019.csv',
 '12-2019.csv',
 'airports.csv']

In [66]:
import wget

ModuleNotFoundError: No module named 'wget'

In [65]:
url = "https://www.python.org/static/img/python-logo@2x.png"
filename = os.path.basename(url)
filename
wget.download(url, 'c:/users/LikeGeeks/downloads/pythonLogo.png')

del url
del filename

ModuleNotFoundError: No module named 'wget'

In [17]:
# Import regular expressions library
import re

In [63]:
list(filter(lambda item: re.fullmatch('\d{2}\-2019\.csv', item, flags=re.I) is not None, os.listdir(data_dir)))

['05-2019.csv',
 '06-2019.csv',
 '07-2019.csv',
 '08-2019.csv',
 '09-2019.csv',
 '10-2019.csv',
 '11-2019.csv',
 '12-2019.csv']

In [64]:
list(filter(lambda item: re.fullmatch('\d{2}\-2019\.csv', item, flags=re.I) is None, os.listdir(data_dir)))

['airports.csv']

In [55]:
[re.fullmatch('^\d{2}\-2019\.csv$', item, flags=re.I) for item in os.listdir(data_dir)][0].string

'05-2019.csv'

In [3]:
primary_source_files = list(filter(
                            lambda item: re.fullmatch(
                                '\d{2}\-2019\.csv',
                                item,
                                flags=re.I
                            ) is not None,
                            os.listdir(data_dir)
                        ))
secondary_source_file = os.listdir(data_dir)[-1]

In [4]:
import pandas as pd

## Primary Data Set

In [5]:
primary_df = pd.concat([
    pd.read_csv(os.path.join(data_dir,filename))
    for filename in primary_source_files
])

primary_df_rows, primary_df_cols = primary_df.shape

print(f"{primary_df_rows:,} rows × {primary_df_cols:,} columns")

5,512,903 rows × 35 columns


In [6]:
primary_df.head()

carrier_code  flight_number origin_airport destination_airport        date  \
0           AS            121            SEA                 ANC  2019-05-01   
1           F9            402            LAX                 DEN  2019-05-01   
2           F9            662            SFO                 DEN  2019-05-01   
3           F9            790            PDX                 DEN  2019-05-01   
4           AS            108            ANC                 SEA  2019-05-01   

   scheduled_elapsed_time tail_number  departure_delay  arrival_delay  \
0                     215      N615AS               -8            -16   
1                     147      N701FR               17             -4   
2                     158      N346FR               44             27   
3                     156      N332FR               24             10   
4                     210      N548AS               -9            -31   

   delay_carrier  ...  HourlyPrecipitation_x  HourlyStationPressure_x  \
0              0  ...                    0.0                    29.59   
1              0  ...                    0.0                    29.65   
2              0  ...                    0.0                    29.98   
3              0  ...                    0.0                    29.98   
4              0  ...                    0.0                    30.18   

   HourlyVisibility_x  HourlyWindSpeed_x     STATION_y  \
0                10.0                8.0  7.027253e+10   
1                10.0                3.0  7.256500e+10   
2                10.0                6.0  7.256500e+10   
3                10.0                0.0  7.256500e+10   
4                10.0                5.0  7.279302e+10   

   HourlyDryBulbTemperature_y  HourlyPrecipitation_y  HourlyStationPressure_y  \
0                        42.0                    0.0                    30.16   
1                        34.0                    0.0                    24.43   
2                        34.0                    0.0                    24.43   
3                        34.0                    0.0                    24.43   
4                        44.0                    0.0                    29.58   

   HourlyVisibility_y HourlyWindSpeed_y  
0                10.0               3.0  
1                 4.0               0.0  
2                 4.0               0.0  
3                 4.0               0.0  
4                10.0               7.0  

[5 rows x 35 columns]

In [7]:
# Get data types and number of null values for each column
pd.concat(
    [
        primary_df.dtypes,
        primary_df.isna().sum()
    ],
    axis=1,
    keys=['data_type','null_count']
)

data_type  null_count
carrier_code                      object           0
flight_number                      int64           0
origin_airport                    object           0
destination_airport               object           0
date                              object           0
scheduled_elapsed_time             int64           0
tail_number                       object       13556
departure_delay                    int64           0
arrival_delay                      int64           0
delay_carrier                      int64           0
delay_weather                      int64           0
delay_national_aviation_system     int64           0
delay_security                     int64           0
delay_late_aircarft_arrival        int64           0
cancelled_code                    object           0
year                               int64           0
month                              int64           0
day                                int64           0
weekday                            int64           0
scheduled_departure_dt            object           0
scheduled_arrival_dt              object           0
actual_departure_dt               object       82715
actual_arrival_dt                 object       88642
STATION_x                        float64        2088
HourlyDryBulbTemperature_x       float64        2088
HourlyPrecipitation_x            float64       10029
HourlyStationPressure_x          float64        2088
HourlyVisibility_x               float64        2088
HourlyWindSpeed_x                float64        2088
STATION_y                        float64        2091
HourlyDryBulbTemperature_y       float64        2091
HourlyPrecipitation_y            float64       10031
HourlyStationPressure_y          float64        2091
HourlyVisibility_y               float64        2091
HourlyWindSpeed_y                float64        2091

**Note:** Eventually, `cancelled_code` will be our target column for a machine-learning algorithm.  
Because the column in the source data is `cancelled_code` and not `canceled_code`, the double-l spelling will be used in this work.

In [8]:
# Combine `year`, `month`, and `day` into a single `string` in the same format as the `date` column
# and check for equality against the actual date column.

# Check only a few rows
(
    primary_df[['year','month','day']][:3]
    .apply(lambda row: '-'.join([val.zfill(2) for val in row.values.astype(str)]), axis=1)
    .equals(
        primary_df.date[:3]
    )
)

# Check all the rows
# (
#     primary_df[['year','month','day']]
#     .apply(lambda row: '-'.join([val.zfill(2) for val in row.values.astype(str)]), axis=1)
#     .equals(
#         primary_df.date
#     )
# )

True

**Note:** The all-rows check, above, is commented out because it takes a long time, but when run, it does show equality between the entire `date` series and the combined `year`-`month`-`day` series.

Because `year`, `month`, and `day` were originally stored as `int64` values, this also tells us that all the values in `date` are properly formatted (no leading or trailing spaces, *etc*.).

The data is therefore redundant, and we don't need both.

`weekday` is likewise redundant, since it can be calculated from `date`.

In [9]:
# Drop redundant date columns
primary_df.drop(
    columns=['year','month','day','weekday'],
    errors='ignore',
    inplace=True
)

primary_df_rows, primary_df_cols = primary_df.shape

print(f"{primary_df_rows:,} rows × {primary_df_cols:,} columns")

5,512,903 rows × 31 columns


In [10]:
# Examine `cancelled_code` column
primary_df.cancelled_code.value_counts()

N    5426150
B      41919
A      23451
C      21370
D         13
Name: cancelled_code, dtype: int64

### What do the codes mean?

According to the United States Department of Transportation Bureau of Transportation Statistics Airlines and Airports data, [Number 14 - On-Time Reporting](https://www.bts.gov/topics/airlines-and-airports/number-14-time-reporting):

**CANCELLATION CODES**
- `A`-Carrier Caused
- `B`-Weather
- `C`-National Aviation System
- `D`-Security

\[`N` is not on the list and represents "None" or "Not cancelled".\]

We are only interested in flights that were cancelled due to weather, so we will keep only rows with `cancelled_code` `B` or `N`.

In [11]:
primary_df = primary_df.loc[primary_df.cancelled_code.isin(['B','N'])]

primary_df_rows, primary_df_cols = primary_df.shape

print(f"{primary_df_rows:,} rows × {primary_df_cols:,} columns")

5,468,069 rows × 31 columns


In [12]:
# Check that there are now only `B` and `N` values
primary_df.cancelled_code.value_counts()

N    5426150
B      41919
Name: cancelled_code, dtype: int64

In [13]:
# Convert `cancelled_code` column into boolean `cancelled` column, where
# `B` = True (*was* cancelled) and `N` = False (*was not* cancelled)

try:
    print("Converting cancelled_code column to boolean… ", end="")
    primary_df.cancelled_code = (primary_df.cancelled_code == 'B')
    primary_df.rename(columns={'cancelled_code':'cancelled'},inplace=True)
    print()
except AttributeError:
    print("Column has already been processed.")

primary_df.cancelled.value_counts()

Converting cancelled_code column to boolean… 


False    5426150
True       41919
Name: cancelled, dtype: int64

In [21]:
# How many flights were cancelled|not cancelled vs. how many departed|arrived

departed = ~primary_df.actual_departure_dt.isna()
arrived = ~primary_df.actual_arrival_dt.isna()

mult_ix = pd.MultiIndex.from_tuples([
    ('departed',True),
    ('departed',False),
    ('arrived',True),
    ('arrived',False),
])

mult_cols = pd.MultiIndex.from_tuples([
    ('cancelled',False),
    ('cancelled',True)
])

pd.DataFrame(
    data=[
        [
            primary_df.loc[(~primary_df.cancelled) & (departed)].shape[0],
            primary_df.loc[(primary_df.cancelled) & (departed)].shape[0]
        ],
        [
            primary_df.loc[(~primary_df.cancelled) & (~departed)].shape[0],
            primary_df.loc[(primary_df.cancelled) & (~departed)].shape[0]
        ],
        [
            primary_df.loc[(~primary_df.cancelled) & (arrived)].shape[0],
            primary_df.loc[(primary_df.cancelled) & (arrived)].shape[0]
        ],
        [
            primary_df.loc[(~primary_df.cancelled) & (~arrived)].shape[0],
            primary_df.loc[(primary_df.cancelled) & (~arrived)].shape[0]
        ]
    ],
    index=mult_ix,
    columns=mult_cols
)

cancelled       
                   False   True
departed True    5426150   1854
         False         0  40065
arrived  True    5424261      0
         False      1889  41919

In [15]:
# Does anything stand out for cancelled flights that still departed?
primary_df.loc[primary_df.cancelled & departed].head().transpose()

16715                17002  \
carrier_code                                     AA                   AA   
flight_number                                  1393                  346   
origin_airport                                  OKC                  DFW   
destination_airport                             DFW                  MSY   
date                                     2019-05-01           2019-05-01   
scheduled_elapsed_time                           69                   85   
tail_number                                  N751UW               N357PV   
departure_delay                                 176                   83   
arrival_delay                                     0                    0   
delay_carrier                                     0                    0   
delay_weather                                     0                    0   
delay_national_aviation_system                    0                    0   
delay_security                                    0                    0   
delay_late_aircarft_arrival                       0                    0   
cancelled                                      True                 True   
scheduled_departure_dt          2019-05-01 18:29:00  2019-05-01 18:25:00   
scheduled_arrival_dt            2019-05-01 19:38:00  2019-05-01 19:50:00   
actual_departure_dt             2019-05-01 21:25:00  2019-05-01 19:48:00   
actual_arrival_dt                               NaN                  NaN   
STATION_x                             72353013967.0        72259003927.0   
HourlyDryBulbTemperature_x                     62.0                 77.0   
HourlyPrecipitation_x                           0.0                  0.0   
HourlyStationPressure_x                       28.46                29.25   
HourlyVisibility_x                             10.0                 10.0   
HourlyWindSpeed_x                               8.0                  7.0   
STATION_y                             72259003927.0        72231012916.0   
HourlyDryBulbTemperature_y                     67.0                 77.0   
HourlyPrecipitation_y                           0.0                  0.0   
HourlyStationPressure_y                       29.29                30.05   
HourlyVisibility_y                             10.0                 10.0   
HourlyWindSpeed_y                              17.0                 11.0   

                                              17815                18640  \
carrier_code                                     AA                   AA   
flight_number                                  2761                 1271   
origin_airport                                  DFW                  IAH   
destination_airport                             STL                  DFW   
date                                     2019-05-01           2019-05-01   
scheduled_elapsed_time                          105                   75   
tail_number                                  N971TW               N898NN   
departure_delay                                 111                  113   
arrival_delay                                     0                    0   
delay_carrier                                     0                    0   
delay_weather                                     0                    0   
delay_national_aviation_system                    0                    0   
delay_security                                    0                    0   
delay_late_aircarft_arrival                       0                    0   
cancelled                                      True                 True   
scheduled_departure_dt          2019-05-01 18:45:00  2019-05-01 19:54:00   
scheduled_arrival_dt            2019-05-01 20:30:00  2019-05-01 21:09:00   
actual_departure_dt             2019-05-01 20:36:00  2019-05-01 21:47:00   
actual_arrival_dt                               NaN                  NaN   
STATION_x                             72259003927.0        72243012960.0   
HourlyDryBulbTemp

In [16]:
# What about non-cancelled flights that didn't arrive?
primary_df.loc[~primary_df.cancelled & ~arrived].head().transpose()

5154                 12535  \
carrier_code                                     AS                   AA   
flight_number                                    55                 2028   
origin_airport                                  SCC                  MEM   
destination_airport                             BRW                  DFW   
date                                     2019-05-01           2019-05-01   
scheduled_elapsed_time                           45                   99   
tail_number                                  N609AS               N749US   
departure_delay                                  29                  398   
arrival_delay                                     0                    0   
delay_carrier                                     0                    0   
delay_weather                                     0                    0   
delay_national_aviation_system                    0                    0   
delay_security                                    0                    0   
delay_late_aircarft_arrival                       0                    0   
cancelled                                     False                False   
scheduled_departure_dt          2019-05-01 10:20:00  2019-05-01 14:56:00   
scheduled_arrival_dt            2019-05-01 11:05:00  2019-05-01 16:35:00   
actual_departure_dt             2019-05-01 10:49:00  2019-05-01 21:34:00   
actual_arrival_dt                               NaN                  NaN   
STATION_x                             70063727406.0        72334013893.0   
HourlyDryBulbTemperature_x                     30.0                 68.0   
HourlyPrecipitation_x                           0.0                  0.0   
HourlyStationPressure_x                       30.27                29.69   
HourlyVisibility_x                              2.0                 10.0   
HourlyWindSpeed_x                              11.0                 11.0   
STATION_y                             70026027502.0        72259003927.0   
HourlyDryBulbTemperature_y                     27.0                 78.0   
HourlyPrecipitation_y                           0.0                  0.0   
HourlyStationPressure_y                       30.29                29.21   
HourlyVisibility_y                              0.5                 10.0   
HourlyWindSpeed_y                              17.0                  9.0   

                                              13657                16277  \
carrier_code                                     WN                   WN   
flight_number                                  2272                 2212   
origin_airport                                  PDX                  ABQ   
destination_airport                             DAL                  DAL   
date                                     2019-05-01           2019-05-01   
scheduled_elapsed_time                          230                  105   
tail_number                                  N931WN               N788SA   
departure_delay                                  -2                   -5   
arrival_delay                                     0                    0   
delay_carrier                                     0                    0   
delay_weather                                     0                    0   
delay_national_aviation_system                    0                    0   
delay_security                                    0                    0   
delay_late_aircarft_arrival                       0                    0   
cancelled                                     False                False   
scheduled_departure_dt          2019-05-01 11:35:00  2019-05-01 16:35:00   
scheduled_arrival_dt            2019-05-01 17:25:00  2019-05-01 19:20:00   
actual_departure_dt             2019-05-01 11:33:00  2019-05-01 16:30:00   
actual_arrival_dt                               NaN                  NaN   
STATION_x                             72698024229.0        72365023050.0   
HourlyDryBulbTemp

In [17]:
primary_obj_cols = primary_df.select_dtypes('object').columns

primary_obj_cols.tolist()

['carrier_code',
 'origin_airport',
 'destination_airport',
 'date',
 'tail_number',
 'scheduled_departure_dt',
 'scheduled_arrival_dt',
 'actual_departure_dt',
 'actual_arrival_dt']

`carrier_code`, `origin_airport`, `destination_airport`, and `tail_number` are legitimate `string`/`text` columns.

`carrier_code` and `tail_number` are for identification purposes, only, though, and so will not be features for the machine-learning model.

`origin_airport` and `destination_airport` will serve as foreign keys to join to the airport data from the secondary dataset.

`date`, `scheduled_departure_dt`, `scheduled_arrival_dt`, `actual_departure_dt`, `actual_arrival_dt` are currently `string`s, but they can be converted to `date`, `datetime` or `timestamp` formats, if necessary prior to uploading to the SQL database.

`actual_departure_dt` and `actual_arrival_dt` can be stored in the database, but they absolutely should ***not*** be used as features for machine learning, as their presence or absence *defines* what it means for a flight to be cancelled, which is exactly what we want the model to predict.

## Secondary Data Set

In [22]:
secondary_df = pd.read_csv(os.path.join(data_dir,secondary_source_file))

secondary_df_rows, secondary_df_cols = secondary_df.shape

print(f"{secondary_df_rows:,} rows × {secondary_df_cols:,} columns")

29,408 rows × 23 columns


In [102]:
# Display dataframe head including all columns
with pd.option_context('display.max_columns', None):
    display(secondary_df.head())

id ident           type  \
0  3632  KLAX  large_airport   
1  3754  KORD  large_airport   
2  3622  KJFK  large_airport   
3  3384  KATL  large_airport   
4  3878  KSFO  large_airport   

                                               name  latitude_deg  \
0                 Los Angeles International Airport     33.942501   
1              Chicago O'Hare International Airport     41.978600   
2              John F Kennedy International Airport     40.639801   
3  Hartsfield Jackson Atlanta International Airport     33.636700   
4               San Francisco International Airport     37.618999   

   longitude_deg  elevation_ft continent   country_name iso_country  \
0    -118.407997         125.0       NaN  United States          US   
1     -87.904800         672.0       NaN  United States          US   
2     -73.778900          13.0       NaN  United States          US   
3     -84.428101        1026.0       NaN  United States          US   
4    -122.375000          13.0       NaN  United States          US   

  region_name iso_region local_region   municipality  scheduled_service  \
0  California      US-CA           CA    Los Angeles                  1   
1    Illinois      US-IL           IL        Chicago                  1   
2    New York      US-NY           NY       New York                  1   
3     Georgia      US-GA           GA        Atlanta                  1   
4  California      US-CA           CA  San Francisco                  1   

  gps_code iata_code local_code  \
0     KLAX       LAX        LAX   
1     KORD       ORD        ORD   
2     KJFK       JFK        JFK   
3     KATL       ATL        ATL   
4     KSFO       SFO        SFO   

                                           home_link  \
0                            https://www.flylax.com/   
1  https://www.flychicago.com/ohare/home/pages/de...   
2                        https://www.jfkairport.com/   
3                    http://www.atlanta-airport.com/   
4                             http://www.flysfo.com/   

                                      wikipedia_link  \
0  https://en.wikipedia.org/wiki/Los_Angeles_Inte...   
1  https://en.wikipedia.org/wiki/O'Hare_Internati...   
2  https://en.wikipedia.org/wiki/John_F._Kennedy_...   
3  https://en.wikipedia.org/wiki/Hartsfield–Jacks...   
4  https://en.wikipedia.org/wiki/San_Francisco_In...   

                                            keywords    score  \
0                                                NaN  1335475   
1                                 CHI, Orchard Place  1503175   
2  Manhattan, New York City, NYC, Idlewild, IDL, ...  1052075   
3                                                NaN  2002475   
4                                           QSF, QBA  1112475   

                last_updated  
0  2020-04-26T22:37:22+00:00  
1  2018-09-16T02:35:35+00:00  
2  2021-11-21T19:43:08+00:00  
3  2018-09-19T14:50:01+00:00  
4  2008-06-13T14:30:04+00:00

In [87]:
# Get data types and number of null values for each column
pd.concat(
    [
        secondary_df.dtypes,
        pd.Series([secondary_df[col].is_unique for col in secondary_df.columns],index=secondary_df.columns),
        secondary_df.isna().sum()
    ],
    axis=1,
    keys=[
        'data_type',
        'unique',
        'null_count'
    ]
)

data_type  unique  null_count
id                    int64    True           0
ident                object    True           0
type                 object   False           0
name                 object   False           0
latitude_deg        float64   False           0
longitude_deg       float64   False           0
elevation_ft        float64   False        2005
continent            object   False       29291
country_name         object   False           0
iso_country          object   False           0
region_name          object   False           0
iso_region           object   False           0
local_region         object   False           0
municipality         object   False          67
scheduled_service     int64   False           0
gps_code             object   False        8670
iata_code            object   False       27378
local_code           object   False        7799
home_link            object   False       28338
wikipedia_link       object   False       26919
keywords             object   False       24451
score                 int64   False           0
last_updated         object   False           0

In [45]:
secondary_df.iata_code.unique().size

2028

Hmm… only 2028 of the `iata_code`s are unique (out of 29,408 total rows).

In [91]:
airport_codes = secondary_df[['iata_code','local_code']].dropna(how='all')

# Get number of null values for each column
pd.DataFrame(
    [
        airport_codes.iata_code.isna().sum(),
        airport_codes.local_code.isna().sum()
    ],
    index=['iata','local'],
    columns=['null_values']
)

iata_code local_code
0           LAX        LAX
1           ORD        ORD
2           JFK        JFK
3           ATL        ATL
4           SFO        SFO
...         ...        ...
29397       NaN       3LA5
29400       NaN       MI51
29404       NaN       FD75
29405       NaN       0OH2
29406       NaN        NM1

[21660 rows x 2 columns]

In [94]:
airport_codes.loc[airport_codes.local_code.isna()]

iata_code local_code
819         CGX        NaN
1982        QWG        NaN
2087        MXG        NaN
2778        NUN        NaN
3175        HLM        NaN
3296        MSD        NaN
3325        DWN        NaN
3340        PMX        NaN
3531        NGZ        NaN
3532        CLG        NaN
5152        AYE        NaN
5153        GVW        NaN
5154        BPA        NaN
5220        BCJ        NaN
5221        CVR        NaN
5377        SYB        NaN
5412        DWS        NaN
5413        BRG        NaN
5429        LHC        NaN
5457        HKB        NaN
5482        UGB        NaN
8137        CKR        NaN
12393       MPE        NaN
16858       TGE        NaN
19579       HBH        NaN
23919       FLU        NaN
24079       AYZ        NaN
24162       ARX        NaN
24420       HCA        NaN
24808       KCG        NaN
24820       CIV        NaN
24858       CLC        NaN
25130       DHB        NaN
25132       DPK        NaN
25302       ESP        NaN
25318       EDA        NaN
25544       FLE        NaN
25563       FSN        NaN
26147       IVH        NaN
26298       KYL        NaN
26499       LRO        NaN
27453       OBK        NaN
27890       MGI        NaN
28036       REH        NaN
28139       FAL        NaN
28306       SFR        NaN
28344       SRF        NaN
28418       SRU        NaN
28950       TWH        NaN
29106       BZS        NaN
29235       SSU        NaN

In [75]:
# See if the two columns match
airport_codes.iata_code.equals(airport_codes.local_code)

False

In [95]:
# What's different?
airport_codes.loc[~airport_codes.iata_code.eq(airport_codes.local_code)]

iata_code local_code
137         SDX        SEZ
140         SCF        SDL
146         NaN        CMA
153         CLD        CRQ
158         HSH        HND
...         ...        ...
29397       NaN       3LA5
29400       NaN       MI51
29404       NaN       FD75
29405       NaN       0OH2
29406       NaN        NM1

[20089 rows x 2 columns]

In [83]:
airport_codes.eq("CLD").any(1)

0        False
1        False
2        False
3        False
4        False
         ...  
25127    False
27475    False
27518    False
28534    False
29396    False
Length: 1979, dtype: bool

In [86]:
airport_codes[airport_codes.eq("LTH").any(axis=1)]

iata_code local_code
24070       LTH        U75

In [58]:
(
    secondary_df
    .loc[~secondary_df.iata_code.isin(secondary_df.iata_code.unique())]
    .sort_values(by='iata_code', ascending=True)
    .head()
    .transpose()
)

Empty DataFrame
Columns: []
Index: [id, ident, type, name, latitude_deg, longitude_deg, elevation_ft, continent, country_name, iso_country, region_name, iso_region, local_region, municipality, scheduled_service, gps_code, iata_code, local_code, home_link, wikipedia_link, keywords, score, last_updated]